pkg/client/interceptor/intercept.go

```diff
@@ -0,0 +1,197 @@
+	"k8s.io/apimachinery/pkg/watch"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type (
+	// GetFn is a function that can be used to intercept GET calls.
+	GetFn func(ctx context.Context, client client.WithWatch, key client.ObjectKey, obj client.Object, opts ...client.GetOption) error
+	// ListFn is a function that can be used to intercept LIST calls.
+	ListFn func(ctx context.Context, client client.WithWatch, list client.ObjectList, opts ...client.ListOption) error
+	// CreateFn is a function that can be used to intercept CREATE calls.
+	CreateFn func(ctx context.Context, client client.WithWatch, obj client.Object, opts ...client.CreateOption) error
+	// DeleteFn is a function that can be used to intercept DELETE calls.
+	DeleteFn func(ctx context.Context, client client.WithWatch, obj client.Object, opts ...client.DeleteOption) error
+	// DeleteAllOfFn is a function that can be used to intercept Collection DELETE calls.
+	DeleteAllOfFn func(ctx context.Context, client client.WithWatch, obj client.Object, opts ...client.DeleteAllOfOption) error
+	// UpdateFn is a function that can be used to intercept UPDATE calls.
+	UpdateFn func(ctx context.Context, client client.WithWatch, obj client.Object, opts ...client.UpdateOption) error
+	// PatchFn is a function that can be used to intercept PATCH calls.
+	PatchFn func(ctx context.Context, client client.WithWatch, obj client.Object, patch client.Patch, opts ...client.PatchOption) error
+	// WatchFn is a function that can be used to intercept WATCH calls.
+	WatchFn func(ctx context.Context, client client.WithWatch, obj client.ObjectList, opts ...client.ListOption) (watch.Interface, error)
+	// SubResourceFn is a function that can be used to intercept SubResource calls.
+	SubResourceFn func(client client.WithWatch, subResource string) client.SubResourceClient
+
+	// SubResourceGetFn is a function that can be used to intercept SubResource GET calls.
+	SubResourceGetFn func(ctx context.Context, client client.SubResourceClient, obj client.Object, subResource client.Object, opts ...client.SubResourceGetOption) error
+	// SubResourceCreateFn is a function that can be used to intercept SubResource CREATE calls.
+	SubResourceCreateFn func(ctx context.Context, client client.SubResourceClient, obj client.Object, subResource client.Object, opts ...client.SubResourceCreateOption) error
+	// SubResourceUpdateFn is a function that can be used to intercept SubResource UPDATE calls.
+	SubResourceUpdateFn func(ctx context.Context, client client.SubResourceClient, obj client.Object, opts ...client.SubResourceUpdateOption) error
+	// SubResourcePatchFn is a function that can be used to intercept SubResource PATCH calls.
+	SubResourcePatchFn func(ctx context.Context, client client.SubResourceClient, obj client.Object, patch client.Patch, opts ...client.SubResourcePatchOption) error
+
+	// Fns contains functions that are called instead of the underlying client's methods.
+	Fns struct {
```

Comment:
Sorry missed this before. Can we maybe replace all Fn with Func and Fns with Funcs? (in all files)

Just reads a lot better in my opinion

Reply:
Done